### PR TITLE
Add safe destroy and reset functions

### PR DESCRIPTION
**Pull request by hishbase**
Add destroy/reset functions to match iOS

destroy - blows away the database and renders the DB unusable

reset - blows away the database but keeps it usable

### CODE REVIEW

File: android/databases/src/main/java/com/coinbase/wallet/libraries/databases/db/Database.kt [hunk 1]
```diff
@@ -186,51 +149,48 @@ class Database<T : RoomDatabaseProvider>() {
...
     @Suppress("UNCHECKED_CAST")
-    inline fun <reified T : DatabaseModelObject> fetchOne(query: String, vararg args: Any): Single<Optional<T>> {
-        return this.fetch<T>(query, *args).map { it.firstOrNull().toOptional() }
-    }
+    inline fun <reified T : DatabaseModelObject> fetchOne(
+        query: String,
+        vararg args: Any
+    ): Single<Optional<T>> = fetch<T>(query, *args)
+        .map { records ->
+            if (records.count() > 1) {
```
Comment (alexshoykhet) [inline android/databases/src/main/java/com/coinbase/wallet/libraries/databases/db/Database.kt]: Good to have this, but I notice we have places in the code where we use `fetchOne` without supplying `LIMIT 1` to the raw query.
Namely: 
```
AddressDao.getAddress(
        blockchain: Blockchain,
        currencyCode: CurrencyCode,
        network: Network,
        address: String
    ): Single<Optional<Address>> {

WalletDao.getWallet(
        blockchain: Blockchain,
        currencyCode: CurrencyCode
    ): Single<Optional<Wallet>> = database.fetchOne(
        "SELECT * FROM Wallet where blockchain = ? AND currencyCode = ?",
        blockchain.rawValue,
        currencyCode.rawValue
    )

ERC20Dao.getERC20(
        contractAddress: String,
        chainId: Int
    ): Single<Optional<ERC20>> = database.fetchOne(
        "SELECT * FROM ERC20 where contractAddress = ? and chainId = ?",
        contractAddress,
        chainId
    )
```
and
```
DappDao.getDapp(
        url: URL
    ): Single<Optional<Dapp>> 
```

Comment (hishbase) [reply android/databases/src/main/java/com/coinbase/wallet/libraries/databases/db/Database.kt]: `limit 1` is not necessary if the lookup is using a unique index. I added this so we can catch instances where we're returning multiple records

File: android/databases/src/main/java/com/coinbase/wallet/libraries/databases/db/Storage.kt [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+ *  @property modelDaos Mapping of database models to Dao. Exposed for inline functions below.
+ */
+class Storage<P : RoomDatabaseProvider> private constructor() {
+    private lateinit var options: StorageOptions
+    private lateinit var provider: RoomDatabaseProvider
+
+    /**
+     * Mapping of class to Observer. Exposed due to inline functions restriction
+     */
+    val observers = ConcurrentHashMap<Class<*>, PublishSubject<*>>()
```
Comment (alexshoykhet) [inline android/databases/src/main/java/com/coinbase/wallet/libraries/databases/db/Storage.kt]: Can you make these `internal` and add `@PublishedApi` annotation to these (inline function magic I just learned)

Comment (hishbase) [reply android/databases/src/main/java/com/coinbase/wallet/libraries/databases/db/Storage.kt]: fixed

**Pull request merged**